\documentclass[12pt, letterpaper]{article}
\usepackage[margin=0.8in]{geometry}
\usepackage[utf8]{inputenc}
\usepackage[numbers]{natbib}
\usepackage{graphicx}
\usepackage{xcolor}
\usepackage{wrapfig}

\usepackage{hyperref}

\title{HMS Summer Institute in Biomedical Informatics}
\author{Mark Keller \thanks{advised by Professor Nils Gehlenborg}}
\date{Summer 2019}

\begin{document}
\maketitle

\begin{abstract}
Visualization...
\end{abstract}

\section{Introduction}
% Background
Following the model of the human genome project which focused resources across institutions to discover a consensus human genome sequence, projects such as ENCODE~\cite{encode2004encode}, GTEx~\cite{lonsdale2013genotype}, the Human Cell Atlas~\cite{regev2017science}, and the 4D Nucleome~\cite{dekker20174d} have been envisioned and executed.
The NIH-sponsored Human Biomolecular Atlas Program (HuBMAP) is using this model to map the human body at the single cell level in a limited number of subjects.
HuBMAP aims to develop spatial mappings of cells and molecules, with new coordinate frameworks that allow querying across levels, from organ to tissue to cell to molecule.
Learning from the struggles of past biomedical and genomic data collection efforts that have developed data portals and visualization tools as afterthoughts, HuBMAP was conceived with the HuBMAP Integration, Visualization, and Engagement (HIVE) group responsible for articulating data access needs from the start~\cite{snyder2019mapping}.


% Importance...

% Current questions

% Existing visualizations tools and browsers

% Existing "static" visualizations in publications

\subsection{Contributions}
% Introduction to Vitessce
Vitessce...

\section{Methods}

\section{Conclusions}
% Brief summary
In this paper...

% Future directions
Future...

\bibliography{main}{}
\bibliographystyle{plainnat}

\end{document}
