\documentclass[12pt, letterpaper]{article}
\usepackage[margin=0.8in]{geometry}
\usepackage[utf8]{inputenc}
\usepackage[numbers]{natbib}
\usepackage{graphicx}
\usepackage{xcolor}
\usepackage{wrapfig}

\usepackage{hyperref}

\title{HMS Summer Institute in Biomedical Informatics}
\author{Mark Keller \thanks{advised by Professor Nils Gehlenborg}}
\date{Summer 2019}

\begin{document}
\maketitle

\begin{abstract}
Abstract here...
\end{abstract}

\section{Introduction}
\subsection{Background}
% Importance of single-cell analysis
Single-cell genomics and transcriptomics enable cell-level measurement of quantities such as gene expression, mutations, and methylation.
Single-cell data reveals cellular heterogeneity and captures the functional states of individual cells that would be hidden by measurements of bulk cell populations that only capture averages~\cite{shapiro2013single}.
Cell types can be teased apart in single-cell data using computational approaches such as dimensionality reduction and clustering based on gene expression profiles~\cite{stegle2015computational,burgess2019spatial}.

% Data generated by HuBMAP
Following the model of the human genome project which focused resources across institutions to discover a consensus human genome sequence, projects such as ENCODE~\cite{encode2004encode}, GTEx~\cite{lonsdale2013genotype}, the Human Cell Atlas~\cite{regev2017science}, and the 4D Nucleome~\cite{dekker20174d} have been envisioned and executed.
The NIH-sponsored Human BioMolecular Atlas Program (HuBMAP) is using this model to map the human body at the single cell level in a limited number of subjects.
HuBMAP aims to develop spatial mappings of cells and molecules, with new coordinate frameworks that allow querying across levels, from organ to tissue to cell to molecule.
Learning from the struggles of past biomedical and genomic data collection efforts that have developed data portals and visualization tools as afterthoughts, HuBMAP was conceived with the HuBMAP Integration, Visualization, and Engagement (HIVE) group responsible for articulating data access needs from the start~\cite{snyder2019mapping}.


% Need for visualization
Visualization of 2D embeddings of cells can facilitate identification of distinct cell populations that group into clusters or subclusters~\cite{wang2017visualization,staahl2016visualization}.
However, such embeddings are dependent on the dimensionality reduction method employed and its parameters.
Often, multiple dimensionality reduction methods are compared, including PCA, t-SNE, and UMAP~\cite{ovchinnikova2019exploring,becht2019dimensionality}.
Technologies that quantify gene expression may also preserve spatial information, for example RNA flourescence \textit{in situ} hybridization (FISH).
In some tissues, it is likely that cells located close to each other are of the same type~\cite{stegle2015computational}.
Spatial information can be visualized as images or as points in a spatial coordinate system, or overlapping layers of different data types aligned to the same coordinate system.

% Introducing Vitessce
Vitessce is a web-based visualization tool for single-cell experiment data, including spatial data from multiple modalities and scatterplot data from arbitrary dimensionality reduction methods.
Visualizations in Vitessce are interactive, with customizable zoom levels, viewer sizes, tooltips, and color encodings.
To display spatial data, including cells and molecules, Vitessce leverages performant geospatial web technologies.
Visualizations within Vitessce have been developed as modular components intended to make them reusable by external applications that require specific functionality provided by Vitessce but focused on different domains or goals.

\subsection{Related work}
% Existing visualizations tools and browsers
The development of web-based visualization tools for spatial single cell transcriptomics is currently very active, in both open- and closed-source models.
Existing open-source and academic tools include Pagoda2~\cite{lake2018integrative}, UCSC Cell Browser~\cite{nowakowski2017spatiotemporal}, the Single Cell Viewer (SCV)~\cite{wang2019single}, SCope~\cite{wouters2019single}, Giotto~\cite{dries2019giotto}, Blue Brain Cell Atlas~\cite{ero2019cell}, and the Allen Cell Explorer (\url{http://allencell.org}).
Several commercial tools are available or have been announced, including those from 10x Genomics and BioTuring.

% Existing "static" visualizations in publications

% Drawbacks of existing tools


\subsection{Contributions}
My work this summer focused on implementing features to support cell selection, cell set management, and linked plot hover events, as well as fixing miscellaneous bugs and ``tech debt''.

% Cell selection, rectangle and polygon
When analyzing clusters of cells and molecules, users may need to make a selection of plot elements to create a set of cells or molecules that fall within a particular area of the plot.
Prior to this summer, a rectangle selection tool had already been implemented in Vitessce.
However, clusters are not typically rectangle-shaped and spatial data of interest can be of any shape.
To address this, one of my first tasks was to develop a polygon selection tool, allowing a user to select points to use as polygon vertices to create a boundary around a potentially irregularly shaped set of cells or other items.
In developing this tool, I used \texttt{turf} (\url{http://turfjs.org}) and \texttt{nebula.gl} (\url{http://nebula.gl/}), JavaScript libraries containing utilities for working with geospatial data.
To integrate these with the Vitessce spatial and scatterplot components, heavy customization of the selection code provided by \texttt{nebula.gl} was required to ensure that the interface was consistent with the rectangular selection tool.
Additionally, using \texttt{nebula.gl} forced us to upgrade the version of one of its peer dependencies, \texttt{deck.gl} (\url{http://deck.gl}), which illuminated bugs that we reported and were quickly fixed by the \texttt{deck.gl} maintainers.
This was my first time reporting bugs in open source software and interacting with external maintainers to help with the bug fix process.

% Cell set management (and set management in general, with independent utility functions)
Once cell sets have been selected, either via one of the selection tools or as a representation of members of clusters resulting from a clustering algorithm, it is useful to be able to organize and update them.
The second main focus of my summer was building a cell set management component to enable users to organize cell sets.
While the development of this component was implemented with cell sets in mind, the component is not specific to cells, and supports sets of any type of item, for example genes or molecules.

% Cell set manager iteration 1: flat
The first iteration of the cell set component allowed users to store sets resulting from interactions with the rectangle and polygon selection tools.
This component supported the operations of naming and renaming sets, viewing set size, and deleting sets.
Additionally, I implemented a feature to store saved sets to the user's web browser to allow them to return to a previous state.
To enable sharing and long-term saving, I implemented features to import and export sets from a JSON file.
This required developing a simple JSON schema to which the imported and exported data would conform.

% Cell set manager iteration 2: hierarchical
While the first iteration of the cell set manager component was useful, through our evaluation of its functionality, we realized that it would be necessary to allow hierarchical organization of sets.
A hierarchy can support organization of sets and subsets, particularly those representing clusters and subclusters at multiple levels of nesting.
This prompted a second iteration of the component, mostly rewriting the code but learning from the flat first implementation.
To display the tree of sets, I explored multiple open source JavaScript UI libraries containing tree layout components: \texttt{ant-design} (\url{http://ant.design/components/tree}), \texttt{blueprint} (\url{http://blueprintjs.com/docs/#core/components/tree}), \texttt{atlaskit tree} (\url{http://atlaskit.atlassian.com/packages/core/tree}).
I ultimately chose to use \texttt{ant-design} because it was the only tree component supporting checking and dragging tree nodes.
As a bonus, the library comes with additional UI components including tabs and popups, and is widely used so will likely continue to be supported for an extended period of time.

In developing the hierarchical cell set manager, I first developed tree and tree node utilities in JavaScript to serve as the internal representation of the hierarchy of sets.
Then, I added UI features that were present in the flat cell set manager, including renaming and deleting sets.
Next, I connected the dragging functionality provided by the tree component to the internal tree representation.
I implemented updated import and export functionalities supporting the hierarchical representation of sets, requiring me to modify the flat import/export JSON schema.

% Cell set manager tabs

% Set operations




% Linked hover
I added support for linked cursor hover events across plots to be able to show contextual information when a user hovers on a particular cell.
On plots peripheral to the currently-hovered plot, emphasis will be placed on the cell over which the cursor rests in the hovered plot.
This allows a user to quickly identify the location of a particular cell across the different visualizations: scatterplots, spatial views, and gene expression heatmaps.

% Tooltips
In addition to highlighting the same cell across plots, hover events allow us to show tooltips containing more information about the cell of interest.
A tooltip is a message positioned near the mouse in a graphical user interface.
In the context of Vitessce, tooltips may contain a cell's ID or list the names of the clusters and subclusters to which it belongs.
To ensure that tooltips do not overflow a particular plot component's boundaries, I implemented collision detection to be able to flip the position of a tooltip relative to the cursor when approaching the edge of a plot.

% Tech debt
% - React hooks
% - CSS -> SCSS
% - More tests

\section{Methods}
% Web-based tools and React
While end users typically do not need to think about the particular technology used to implement a visualization tool, it is something I thought about this summer.
Web technologies currently provide a good balance of accessibility, performance, and development experience for visualization, and will continue to do so in the near future.
Web browsers support time-tested visualization technologies such as SVG, WebGL, and the canvas element, while improving the development experience with recent standards like WebAssembly and modern JavaScript language features.
The community around open-source JavaScript libraries is large and very active.
Vitessce is implemented using the React JavaScript framework originally developed for Facebook but now distributed as open source (\url{https://reactjs.org}).
React is a framework that simplifies the development of modular, component-based web applications.

% Cartographic technologies
Vitessce performs spatial visualization by leveraging libraries developed with a focus on cartography and mapping.
The open source Uber-maintained Deck.gl JavaScript library is powerful because it implements reactive updates that are similar to those used by React.
Deck.gl performs diffing on the data passed to each view layer to detect updates and invalidate the current state, just as React performs diffing on the virtual DOM in response to data changes to determine when to update the browser's DOM.

Often, it makes sense to render contextual elements in the DOM rather than in a visualization layer, whether for performance reasons or to keep the visualization clean to in preparation for downloading.
To synchronize HTML element positions with positions of data points in visualization layers, we perform projections from the data coordinate space to the browser coordinate space.
This is how I chose to implement the cell-hover emphasis and tooltip functionalities.

% Component re-use
The development of Vitessce has focused on separating logic into components that can operate independently to enable them to be imported and re-used by other projects, including HuBMAP data portals and tissue viewers.
To achieve this goal, Vitessce does not maintain a global state that is used by child components.
Instead, there is an event-based mechanism, with wrapper subscriber components that pass state down to children along with publisher update functions.
This allows external usage of components to be done by setting up custom publishing and subscription wrappers rather than using the Vitessce-specific wrappers.
A drawback of this approach is that events are asynchronous, so implementing a history mechanism may prove to be difficult, as there is no one source of truth for events.


\section{Conclusions}
% Brief summary
Conclusions here...

% Questions that Vitessce can potentially answer

% Future directions
% - more customization options
% - layout preferences persistence
% - database for cell sets
% - image processing pipeline
% - more interactive heatmap
Future directions here...

\subsection{Reflection}

% What I got out of this summer
% - HuBMAP conference experience
% - more team-based research and software development
% - a focus on visualization
% - grad school application process



\bibliography{main}{}
\bibliographystyle{unsrt}

\end{document}
