\documentclass[12pt, letterpaper]{article}
\usepackage[margin=0.8in]{geometry}
\usepackage[utf8]{inputenc}
\usepackage[numbers]{natbib}
\usepackage{graphicx}
\usepackage{xcolor}
\usepackage{wrapfig}
\usepackage{hyperref}

\title{HMS Summer Institute in Biomedical Informatics}
\author{Mark Keller \thanks{advised by Professor Nils Gehlenborg}}
\date{Summer 2019}

\begin{document}
\maketitle

\begin{abstract}
Abstract here...
\end{abstract}

\section{Introduction}
\subsection{Background}
% Importance of single-cell analysis
Single-cell genomics and transcriptomics enable cell-level measurement of quantities such as gene expression, mutations, and methylation.
Single-cell data reveals cellular heterogeneity and captures the functional states of individual cells that would be hidden by measurements of bulk cell populations that only capture averages~\cite{shapiro2013single}.
Cell types can be teased apart in single-cell data using computational approaches such as dimensionality reduction and clustering based on gene expression profiles~\cite{stegle2015computational,burgess2019spatial}.

% Data generated by HuBMAP
Following the model of the human genome project which focused resources across institutions to discover a consensus human genome sequence, projects such as ENCODE~\cite{encode2004encode}, GTEx~\cite{lonsdale2013genotype}, the Human Cell Atlas~\cite{regev2017science}, and the 4D Nucleome~\cite{dekker20174d} have been envisioned and executed.
The NIH-sponsored Human BioMolecular Atlas Program (HuBMAP) is using this model to map the human body at the single cell level in a limited number of subjects.
HuBMAP aims to develop spatial mappings of cells and molecules, with new coordinate frameworks that allow querying across levels, from organ to tissue to cell to molecule.
Learning from the struggles of past biomedical and genomic data collection efforts that have developed data portals and visualization tools as afterthoughts, HuBMAP was conceived with the HuBMAP Integration, Visualization, and Engagement (HIVE) group responsible for articulating data access needs from the start~\cite{snyder2019mapping}.

% Need for visualization
Visualization of 2D embeddings of cells can facilitate identification of distinct cell populations that group into clusters or subclusters~\cite{wang2017visualization,staahl2016visualization}.
However, such embeddings are dependent on the dimensionality reduction method employed and its parameters.
Often, multiple dimensionality reduction methods are compared, including PCA, t-SNE, and UMAP~\cite{ovchinnikova2019exploring,becht2019dimensionality}.
Technologies that quantify gene expression may also preserve spatial information, for example RNA florescence \textit{in situ} hybridization (FISH).
In some tissues, it is likely that cells located close to each other are of the same type~\cite{stegle2015computational}.
Spatial information can be visualized as images or as points in a spatial coordinate system, or overlapping layers of different data types aligned to the same coordinate system.

% Introducing Vitessce
Vitessce (\url{http://vitessce.io}) is a web-based visualization tool for single-cell experiment data, including spatial data from multiple modalities and scatterplot data from arbitrary dimensionality reduction methods.
Visualizations in Vitessce are interactive, with customizable zoom levels, viewer sizes, tooltips, and color encodings.
To display spatial data, including cells and molecules, Vitessce leverages performant geospatial web technologies.
Visualizations within Vitessce have been developed as modular components intended to make them reusable by external applications that require specific functionality provided by Vitessce but focused on different domains or goals.

\subsection{Related work}
% Existing visualizations tools and browsers
The development of web-based visualization tools for spatial single cell transcriptomics is currently very active, in both open- and closed-source models.
Existing open-source and academic tools include Pagoda2~\cite{lake2018integrative}, UCSC Cell Browser~\cite{nowakowski2017spatiotemporal}, the Single Cell Viewer~\cite{wang2019single}, SCope~\cite{wouters2019single}, Giotto~\cite{dries2019giotto}, Blue Brain Cell Atlas~\cite{ero2019cell}, Allen Cell Explorer \footnote{\url{https://allencell.org}}, and cellxgene \footnote{\url{https://chanzuckerberg.github.io/cellxgene/}}.
Several commercial tools are available or have been announced, including those from 10x Genomics and BioTuring.
% Existing "static" visualizations in publications
Tools for static visualization of single-cell data are also popular, with use cases including creating publication-ready figures or fitting into traditional analysis workflows.

% Drawbacks of existing tools
Many of these existing tools focus on a particular aspect of single-cell data visualization, but none support the multi-view linked spatial, scatterplot, and heatmap views that Vitessce supports.
Additionally, the interfaces of tools like Single Cell Viewer are not user-friendly, requiring users to scroll and change pages many times to view different aspects of their data.
Vitessce focuses on user experience to allow customization and arbitrary simultaneous views of data, while also recognizing that certain use cases may be out of scope and better accommodated by individual Vitessce components rather than the app as a whole.

\subsection{Contributions}
My work this summer focused on implementing features to support cell selection, cell set management, and linked plot hover events, as well as fixing miscellaneous bugs and ``tech debt''.

% Cell selection, rectangle and polygon
When analyzing clusters of cells and molecules, users may need to make a selection of plot elements to create a set of cells or molecules that fall within a particular area of the plot.
Prior to this summer, a rectangle selection tool had already been implemented in Vitessce.
However, clusters are not typically rectangle-shaped and spatial data of interest can be of any shape.
To address this, one of my first tasks was to develop a polygon selection tool, allowing a user to select points to use as polygon vertices to create a boundary around a potentially irregularly shaped set of cells or other items.
In developing this tool, I used \texttt{turf} \footnote{\url{http://turfjs.org}} and \texttt{nebula.gl} \footnote{\url{http://nebula.gl/}}, JavaScript libraries containing utilities for working with geospatial data.
To integrate these with the Vitessce spatial and scatterplot components, heavy customization of the selection code provided by \texttt{nebula.gl} was required to ensure that the interface was consistent with the rectangular selection tool.
Additionally, using \texttt{nebula.gl} forced us to upgrade the version of one of its peer dependencies, \texttt{deck.gl} \footnote{\url{http://deck.gl}}, which illuminated bugs that we reported and were quickly fixed by the \texttt{deck.gl} maintainers.
This was my first time reporting bugs in open source software and interacting with external maintainers to help with the bug fix process.

% Cell set management (and set management in general, with independent utility functions)
Once cell sets have been selected, either via one of the selection tools or as a representation of members of clusters resulting from a clustering algorithm, it is useful to be able to organize and update them.
The second main focus of my summer was building a cell set management component to enable users to organize cell sets.
While this component motivated by cell sets, organization of sets is important for other types of items, so the component was developed at an abstract level to support other set types, for instance genes or molecules.

% Cell set manager mock ups
Prior to writing any code for the set management component, I first created UI mockups in Sketch, a UI design and prototyping application, in order to get feedback regarding the features and layout that the component would support.
Through this process, we aimed to get clarification about the desired interactions or workflow for managing cell sets, which was successful prompted me to develop the component based on the agreed-upon mockup.

% Cell set manager iteration 1: flat
The first iteration of the cell set component allowed users to store sets resulting from interactions with the rectangle and polygon selection tools.
This component supported the operations of naming and renaming sets, viewing set size, and deleting sets.
Additionally, I implemented a feature to store saved sets to the user's web browser to allow them to return to a previous state.
To enable sharing and long-term saving, I implemented features to import and export sets from a JSON file.
This required developing a simple JSON schema to which the imported and exported data would conform.

% Cell set manager iteration 2: hierarchical
While the first iteration of the cell set manager component was useful, through our evaluation of its functionality, we realized that it would be necessary to allow hierarchical organization of sets.
A hierarchy can support organization of sets and subsets, particularly those representing clusters and subclusters at multiple levels of nesting.
This prompted a second iteration of the component, mostly rewriting the code but learning from the flat first implementation.
To display the tree of sets, I explored multiple open source JavaScript UI libraries containing tree layout components: \texttt{ant-design} \footnote{\url{https://ant.design/components/tree/}}, \texttt{blueprint} \footnote{\url{https://blueprintjs.com/docs/\#core/components/tree}}, \texttt{atlaskit tree} \footnote{\url{https://atlaskit.atlassian.com/packages/core/tree}}.
I ultimately chose to use \texttt{ant-design} because it was the only tree component supporting checking and dragging tree nodes.
As a bonus, the library comes with additional UI components including tabs and popups, and is widely used so will likely continue to be supported for an extended period of time.

In developing the hierarchical cell set manager, I first developed tree and tree node utilities in JavaScript to serve as the internal representation of the hierarchy of sets.
I added UI features that were present in the flat cell set manager, including renaming and deleting sets, and connected the dragging functionality provided by the tree component to the internal tree representation.
To allow coloring of sets, I added a color picker to each tree node, triggering an event which updates the scatterplot and spatial plots upon color change.
Cell sets can be viewed using a ``view'' popup menu and choosing the current set or a particular level of its descendants.
This menu is dynamic to accommodate tree nodes at different levels.
These view options allow multiple cell sets at the same level of the hierarchy to be viewed simultaneously, distinguished by their differing colors.
For example, if the clusters discovered by a particular clustering algorithm are stored under the same parent node, the visualization of all clusters can be done by selecting the parent node and choosing to view the 1st level descendants.

% Cell set manager tabular import/export
I implemented updated import and export functionalities supporting the hierarchical representation of sets, requiring me to modify the flat import/export JSON schema.
Another request we received to make the importing and exporting more user-friendly was to add support for a tabular file format in addition to the JSON format.
I defined a tabular format that encodes the hierarchy implicitly and uses the concept of tidy data to define cells by rows and set attributes as column values~\cite{wickham2014tidy}.
Tabular files are more straightforward for researchers without computational experience, as they can be created in spreadsheet software or with built-in R functions.

% Cell set manager tabs
Within the cell set manager, I also implemented a tab interface to ease the management of deeply nested tree nodes.
From the view menu of a tree node, there is an option to open the node as a new tab, in which the children of the node of interest are displayed as the top level of the hierarchy, simplifying the interface.
These tabs act like web browser tabs that a user would expect, and can be viewed and closed in the same way.

% Set operations
It is useful to be able to derive new sets from the existing sets that may have been the result of a selection interaction or a clustering algorithm.
Another part of the cell set manager component that I implemented this summer is the set operations interface.
This interface supports taking unions, intersections, and complements that become new sets.
One example of the usage of these operations in the context of cell sets would be to manually create a new set of all neurons from multiple sets representing different neuron types.
These operations can be performed on multiple operands by using the checkboxes corresponding to each tree node.

% Linked hover
I added support for linked cursor hover events across plots to be able to show contextual information when a user hovers on a particular cell.
On plots peripheral to the currently-hovered plot, emphasis will be placed on the cell over which the cursor rests in the hovered plot.
This allows a user to quickly identify the location of a particular cell across the different visualizations: scatterplots, spatial views, and gene expression heatmaps.

% Tooltips
In addition to highlighting the same cell across plots, hover events allow us to show tooltips containing more information about the cell of interest.
A tooltip is a message positioned near the mouse in a graphical user interface.
In the context of Vitessce, tooltips may contain a cell's ID or list the names of the clusters and subclusters to which it belongs.
To ensure that tooltips do not overflow a particular plot component's boundaries, I implemented collision detection to be able to flip the position of a tooltip relative to the cursor when approaching the edge of a plot.

My implementation of tooltips and hover emphasis markings was done by rendering contextual elements in the DOM rather than in a visualization layer directly, both performance reasons and to keep the visualization clean to in preparation for potential downloading.
To synchronize HTML element positions with positions of data points in visualization layers, we perform projections from the data coordinate space to the browser coordinate space.
This approach also follows the decoupling paradigms followed throughout Vitessce, making the extra elements modular and composable.

% Tech debt
% - React hooks
Along with the development of individual features for Vitessce, I also helped to reduce tech debt and streamline some aspects of the code base.
This included adhering to the recent additions to the React library documentation regarding the implementation of new components as JavaScript functions rather than JavaScript classes.
However, because pure JavaScript functions do not allow for the storage of component state variables (as classes do), to ``hook into'' React's state functionality, usage of additional functions called ``hooks'' is necessary.
React hooks can make code more readable by allowing code blocks corresponding to a particular task to be grouped together, while class-based components require separation of code based on component lifecycle events.
% - CSS -> SCSS
I migrated our code for styling Vitessce from CSS to Sass \footnote{\url{http://sass-lang.com}}, allowing us to leverage CSS preprocessing for nesting styles, computed variables, and style reuse via mixins.
% - More tests
In the development of the cell set manager, I wrote unit tests for each of its functions, as well as tests for the import/export schema validation.
% - Protobuf
My final task for the summer was exploring a data serialization tool called Protobuf \footnote{\url{http://developers.google.com/protocol-buffers}} developed by Google for data input and output in a language-agnostic way.
Improved serialization and compression could improve loading times of large datasets in Vitessce.
The first part of this exploration process entailed creating a small pipeline to make a full trip: from a Protobuf schema to a Python data structure to a serialized data structure sent the browser to be decoded into a JSON data structure.

These updates are invisible to Vitessce users but will help future developers to easily understand, add, and update functionality in Vitessce.


\section{Discussion}
It has been a fun experience to spend the summer immersed in a lab focused more on methods and infrastructure development than hypothesis-based research.
While I enjoy traditional software development, visualization research software and infrastructure development is particularly exciting to me, as it moves fast and serves as a critical part of the translational, educational, and exploratory functions of research.
The field of life sciences research software development is currently struggling with the problems of reproducibility and long-term maintenance, issues important to the adoption of such software~\cite{gruning2018practical,belmann2015bioboxes}.
% Adoption of software development industry techniques in research
Software development practices in academia seem to lag far behind those in industry, possibly due to differences in incentives.
For instance, the concept of ``continuous integration'' has existed since 1991, yet articles as recently as 2017 were still treating it as a foreign concept to life sciences researchers~\cite{booch1991object,beaulieu2017reproducibility}.
Continuous integration and testing require more time and effort up front but prevent software from regressing and decrease the cognitive overhead for maintenance later.
It is promising that projects like Vitessce, HiGlass~\cite{kerpedjiev2018higlass}, and others in the lab are committed to these best practices.

% Web-based tools and React
Vitessce adopts both general UI technologies and specific geospatial technologies for the single-cell visualization domain.
While end users typically do not need to think about the particular technology used to implement a visualization tool, it is something I thought about this summer.
Web technologies currently provide a good balance of accessibility, performance, and development experience for visualization, and will continue to do so in the near future.
Web browsers support time-tested visualization technologies such as SVG, WebGL, and the canvas element, while improving the development experience with recent standards like WebAssembly and modern JavaScript language features.
The community around open-source JavaScript libraries is large and very active.
Vitessce is implemented using the React JavaScript framework originally developed for Facebook but now distributed as open source \footnote{\url{https://reactjs.org}}.
React is a framework that simplifies the development of modular, component-based web applications.

% Cartographic technologies
Vitessce performs spatial visualization by leveraging libraries developed with a focus on cartography and mapping.
The open source Uber-maintained Deck.gl JavaScript library is powerful because it implements reactive updates that are similar to those used by React.
Deck.gl performs diffing on the data passed to each view layer to detect updates and invalidate the current state, just as React performs diffing on the virtual DOM in response to data changes to determine when to update the browser's DOM.

% Component re-use
The development of Vitessce has focused on separating logic into components that can operate independently to enable them to be imported and re-used by other projects, including HuBMAP data portals and tissue viewers.
To achieve this goal, Vitessce does not maintain a global state that is used by child components.
Instead, there is an event-based mechanism, with wrapper subscriber components that pass state down to children along with publisher update functions.
This allows external usage of components to be done by setting up custom publishing and subscription wrappers rather than using the Vitessce-specific wrappers.
A drawback of this approach is that events are asynchronous, so implementing a history mechanism may prove to be difficult, as there is no one source of truth for events.

% Packaging CSS/JS for distribution on npm
To help others to adopt open source software, packaging and distribution are important to think about.
In the case of the Vitessce components, we transpile the JavaScript code and distribute it on NPM.
The CSS code for styling the components must also be shared along with the JavaScript code.
Because of my work transitioning to a more modular styling system using SCSS files, we can distribute modular stylesheets for users that are using only a specific component or a small set of components.

\section{Conclusions}
% Questions that Vitessce can potentially answer
It is hypothesized that analysis of single cell data in healthy human tissue will provide insights into the mechanisms of cell differentiation and other cellular processes.
Better understandings of mechanisms at the cellular level and beyond will impact diagnosis and treatment of human disease, for example by informing testing procedures or drug development efforts.
% Brief summary
New capabilities are being implemented in Vitessce at the same time as new single cell spatial transcriptomics data sources and data types are becoming available.
Vitessce will make exploration of single cell data across different embeddings and visualization approaches accessible to both the casual and seasoned analyst of single cell data.
Accessibility will come not only through the availability of the web, but also through the user-centered feature development process, enabled by constant feedback from experts in general biomedical visualization and single cell data analysis.
Its positioning within the scope of HuBMAP will enforce the adoption of current data and visualization standards, and will provide the resources to define new standards where applicable.

% Future directions
% - more customization options
% - layout preferences persistence
% - database for cell sets
% - image processing pipeline
% - more interactive heatmap
Features planned for the next phase of development on Vitessce involve user interface customization and image processing.
To allow users to return to the layout configuration of a previous browser session, layout settings need to be persisted to the local browser storage and loaded upon visiting the web application.
Besides layout, making other customization options available to the user such as enabling and disabling visibility of tooltips and persisting these choices will be important.
Refinement of the image processing pipeline is another planned development focus, during which support for additional file types and multi-channel images will be added.
Experiments with other image viewers and tiling methods, including HiGlass~\cite{kerpedjiev2018higlass}, are planned to ensure that the optimal viewing method is employed.
Further interactivity features are planned, including implementation of cell and gene selections via the heatmap component to complement those already available in the spatial and scatterplot components.


\subsection{Reflection}

% What I got out of this summer
% - HuBMAP conference experience
% - more team-based research and software development
% - a focus on visualization
% - grad school application process

\subsection*{Acknowledgements}
I am grateful to the Summer Institute and the Gehlenborg Lab for welcoming me this summer, and especially to Chuck for mentorship regarding development of Vitessce and to Dr. Gehlenborg for mentorship regarding Vitessce as well as visualization research in general and the graduate school application process.


\bibliography{main}{}
\bibliographystyle{unsrt}

\end{document}
